runtime/jvmti/jvmtiStackFrame.c

```diff
@@ -217,29 +217,50 @@ jvmtiGetThreadListStackTraces(jvmtiEnv* env,
 		vm->internalVMFunctions->acquireExclusiveVMAccess(currentThread);
 
 		stackInfo = j9mem_allocate_memory(((sizeof(jvmtiStackInfo) + (max_frame_count * sizeof(jvmtiFrameInfo))) * thread_count) + sizeof(jlocation), J9MEM_CATEGORY_JVMTI_ALLOCATE);
-		if (stackInfo == NULL) {
+		if (NULL == stackInfo) {
 			rc = JVMTI_ERROR_OUT_OF_MEMORY;
 		} else {
-			jvmtiFrameInfo * currentFrameInfo = (jvmtiFrameInfo *) ((((UDATA) (stackInfo + thread_count)) + sizeof(jlocation)) & ~sizeof(jlocation));
-			jvmtiStackInfo * currentStackInfo = stackInfo;
+			jvmtiFrameInfo *currentFrameInfo = (jvmtiFrameInfo *)((((UDATA) (stackInfo + thread_count)) + sizeof(jlocation)) & ~sizeof(jlocation));
+			jvmtiStackInfo *currentStackInfo = stackInfo;
 
-			while (thread_count != 0) {
+			while (0 != thread_count) {
 				jthread thread = *thread_list;
-				J9VMThread * targetThread;
-				j9object_t threadObject;
+				J9VMThread *targetThread = NULL;
+				j9object_t threadObject = NULL;
+#if JAVA_SPEC_VERSION >= 19
+				j9object_t continuationObject = NULL;
+#endif /* JAVA_SPEC_VERSION >= 19 */
 
 				if (thread == NULL) {
 					rc = JVMTI_ERROR_NULL_POINTER;
 					goto deallocate;
 				}
 
-				if (!isSameOrSuperClassOf(J9VMJAVALANGTHREAD_OR_NULL(vm), J9OBJECT_CLAZZ(currentThread, *((j9object_t *) thread)))) {
+				if (!isSameOrSuperClassOf(J9VMJAVALANGTHREAD_OR_NULL(vm), J9OBJECT_CLAZZ(currentThread, *((j9object_t *)thread)))) {
 					rc = JVMTI_ERROR_INVALID_THREAD;
 					goto deallocate;
 				}
 
-				threadObject = *((j9object_t*) thread);
-				targetThread = J9VMJAVALANGTHREAD_THREADREF(currentThread, threadObject);
+				threadObject = J9_JNI_UNWRAP_REFERENCE(thread);
+#if JAVA_SPEC_VERSION >= 19
+				if (IS_VIRTUAL_THREAD(currentThread, threadObject)) {
+					j9object_t carrierThread = (j9object_t)J9VMJAVALANGVIRTUALTHREAD_CARRIERTHREAD(currentThread, threadObject);
+					if (NULL != carrierThread) {
+						targetThread = J9VMJAVALANGTHREAD_THREADREF(currentThread, carrierThread);
+					} else {
+						continuationObject = (j9object_t)J9VMJAVALANGVIRTUALTHREAD_CONT(currentThread, threadObject);
+					}
+				} else
+#endif /* JAVA_SPEC_VERSION >= 19 */
+				{
+					targetThread = J9VMJAVALANGTHREAD_THREADREF(currentThread, threadObject);
+#if JAVA_SPEC_VERSION >= 19
+					if ((NULL != targetThread) && (targetThread->threadObject != threadObject)) {
+						continuationObject = (j9object_t)J9VMJAVALANGVIRTUALTHREAD_CONT(currentThread, targetThread->threadObject);
+						targetThread = NULL;
+					}
+#endif /* JAVA_SPEC_VERSION >= 19 */
```

Comment:
small suggestion, could the two ifdef blocks be merged using?
```c++
#ifdef
if (IS_VIRTUAL_THREAD(currentThread, threadObject)) {
    if (NULL != carrierThread) {
        threadObject = carrierThread;
    } else {
        continuationObject = (j9object_t)J9VMJAVALANGVIRTUALTHREAD_CONT(currentThread, threadObject);
    }
#endif
```
as threadref on a virtualthread will always be null, don't need to explicitly set it

Reply:
don't quite understand this

Reply:
please ignore above comment,

on second thought, this will not guarantee correctness as exclusive access do not ensure vthread transition completes, meaning it could halt thread during transition. In such cases we will need to check against `vmThread->currentContinuation` to confirm if a vthread( or carrier) is mounted or not.

I will post a pseudo code shortly.

@babsingh any objections or comments?

Reply:
- Can we encapsulate all changes in `jvmtiInternalGetStackTrace` by passing in the `threadObject`?
- Do we need to correct all places where `jvmtiInternalGetStackTrace` is used?

Reply:
> * Can we encapsulate all changes in `jvmtiInternalGetStackTrace` by passing in the `threadObject` to `jvmtiInternalGetStackTrace`?

No, this is only an issue under `exclusiveVMAccess` since exclusive doesn't respect the inspector count. so we can be stopped inside the begin/end scope

> * Do we need to correct all places where `jvmtiInternalGetStackTrace` is used?

only when we do not use `getVMThread` to ensure thread sync